platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/DefaultProblemBuilder.java

```diff
@@ -39,17 +39,17 @@ public class DefaultProblemBuilder implements InternalProblemBuilder {
     private final Map<String, Object> additionalData;
     private boolean collectLocation = false;
 
-    public DefaultProblemBuilder(Problem problem) {
-        this.label = problem.getLabel();
-        this.category = problem.getCategory();
-        this.severity = problem.getSeverity();
-        this.locations = new ArrayList<ProblemLocation>(problem.getLocations());
-        this.details = problem.getDetails();
-        this.docLink = problem.getDocumentationLink();
-        this.solutions = new ArrayList<String>(problem.getSolutions());
-        this.exception = problem.getException();
-        this.additionalData = new HashMap<String, Object>(problem.getAdditionalData());
-        this.namespace = problem.getCategory().getNamespace();
+    public DefaultProblemBuilder(ProblemReport problem) {
+        this.label = problem.getProblemDefinition().getLabel();
+        this.category = problem.getProblemDefinition().getCategory();
+        this.severity = problem.getProblemDefinition().getSeverity();
+        this.locations = new ArrayList<ProblemLocation>(problem.getProblemContext().getLocations());
+        this.details = problem.getProblemContext().getDetails();
+        this.docLink = problem.getProblemDefinition().getDocumentationLink();
+        this.solutions = new ArrayList<String>(problem.getProblemDefinition().getSolutions());
+        this.exception = problem.getProblemContext().getException();
+        this.additionalData = new HashMap<String, Object>(problem.getProblemContext().getAdditionalData());
+        this.namespace = problem.getProblemDefinition().getCategory().getNamespace();
```

Comment:
🤔 Maybe instead of wrapping/unwrapping the definition and context objects, we should use them as containers?

The answer is not as simple here for me, as we would need to make both classes mutable, but package-level constructors and setter methods could do all of these.

Reply:
I'll also check how can I improve the situation here.

Reply:
> as we would need to make both classes mutable

That, we should not do, the ProblemReport class has to be immutable.

Reply:
Refactoring the builders is a bit more involved. I propose we tackle that in a separate PR.

Reply:
No problem, this is a back-end thing we can do anytime.